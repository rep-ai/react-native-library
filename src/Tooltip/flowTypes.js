// @flow
export type TooltipProps = {|
    description: string,
    trianglePosition: 'top' | 'right' | 'bottom' | 'left' | 'none',
    visible?: bool,
    title?: string,
    onClose?: () => void,
|}
