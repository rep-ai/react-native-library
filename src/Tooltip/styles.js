// @flow
import { StyleSheet } from 'react-native'

import { elevationToShadowProps } from '../util'

const elevation = 24
export const triangleSize = 20
export const triangleVisualOffset = 1
export const triangleBorderRadius = 3

export default StyleSheet.create({
    innerContainer: {
        elevation,
        ...elevationToShadowProps(elevation),
        backgroundColor: 'white',
        borderRadius: 10,
        padding: 36,
        alignItems: 'center',
        justifyContent: 'center',
        maxHeight: '70%',
        maxWidth: '80%',
    },

    container: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
    },

    text: {
        fontSize: 21,
        textAlign: 'center',
        backgroundColor: 'transparent',
        color: '#444',
    },

    title: {
        fontSize: 32,
        marginBottom: 18,
    },

    triangle: {
        position: 'absolute',
        width: triangleSize * 2,
        height: triangleSize * 2,
        backgroundColor: 'white',
        transform: [{ rotate: '45deg' }],
    },

    top: {
        top: -triangleSize,
        borderTopLeftRadius: triangleBorderRadius,
    },

    bottom: {
        bottom: -triangleSize,
        borderBottomRightRadius: triangleBorderRadius,
    },

    left: {
        left: -triangleSize,
        borderBottomLeftRadius: triangleBorderRadius,
    },

    right: {
        right: -triangleSize,
        borderTopRightRadius: triangleBorderRadius,
    },

    none: {},
})
