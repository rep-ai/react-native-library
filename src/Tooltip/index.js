// @flow
import React, { PureComponent } from 'react'
import { View, Text } from 'react-native'

import ModalScreen from '../ModalScreen'
import capitalize from '../UtilityMethods/capitalize'
import styles, { triangleSize } from './styles'

type TooltipProps = {|
    visible: bool,
    offset?: number,
    title?: string,
    description: string,
    trianglePosition: 'top' | 'right' | 'bottom' | 'left' | 'none',
    onClose?: () => void,
    style?: *,
|}

class Tooltip extends PureComponent<TooltipProps> {
    static defaultProps = {
        visible: true,
    }

    render() {
        const {
            visible,
            title,
            description,
            onClose,
            trianglePosition,
            style,
            offset,
        } = this.props

        return (
            <ModalScreen
                style={[styles.container].concat(style)}
                touchingBackgroundShouldHide={true}
                visible={visible}
                onRequestClose={onClose}>
                <View style={[styles.innerContainer, trianglePosition === 'none' ? {} : {
                    [`margin${capitalize(trianglePosition)}`]: Number(offset) + triangleSize,
                }]}>
                    <View style={[styles.triangle, styles[trianglePosition]]} />
                    {!!title && <Text style={[styles.text, styles.title]}>{title}</Text>}
                    <Text style={styles.text}>{description}</Text>
                </View>
            </ModalScreen>
        )
    }
}

export default Tooltip
