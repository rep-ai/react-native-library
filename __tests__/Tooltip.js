import 'react-native'
import React from 'react'
import Tooltip from '../src/Tooltip'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'
const text = `Lorem Ipsum is simply dummy text of the printing and typesetting industry.`

it('renders correctly', () => {
    // NOTE: Bugged test.
    ['top', 'left', 'bottom', 'right'].map((trianglePosition) => {
        const tree = renderer.create(
            <Tooltip
                description={text}
                trianglePosition={trianglePosition}
            />
        )

        expect(tree.toJSON()).toMatchSnapshot()

        return null
    })
})
