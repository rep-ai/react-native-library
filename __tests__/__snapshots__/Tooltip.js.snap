// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    collapsable={undefined}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 1334,
          },
        ],
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID={undefined}
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 10,
              "elevation": 24,
              "justifyContent": "center",
              "maxHeight": "70%",
              "maxWidth": "80%",
              "padding": 36,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 14.399999999999999,
              },
              "shadowOpacity": 0.216,
              "shadowRadius": 12.96,
            },
            Object {
              "marginTop": NaN,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "height": 40,
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotate": "45deg",
                  },
                ],
                "width": 40,
              },
              Object {
                "borderTopLeftRadius": 3,
                "top": -20,
              },
            ]
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "backgroundColor": "transparent",
              "color": "#444",
              "fontSize": 21,
              "textAlign": "center",
            }
          }
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry.
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly 2`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    collapsable={undefined}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 1334,
          },
        ],
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID={undefined}
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 10,
              "elevation": 24,
              "justifyContent": "center",
              "maxHeight": "70%",
              "maxWidth": "80%",
              "padding": 36,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 14.399999999999999,
              },
              "shadowOpacity": 0.216,
              "shadowRadius": 12.96,
            },
            Object {
              "marginLeft": NaN,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "height": 40,
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotate": "45deg",
                  },
                ],
                "width": 40,
              },
              Object {
                "borderBottomLeftRadius": 3,
                "left": -20,
              },
            ]
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "backgroundColor": "transparent",
              "color": "#444",
              "fontSize": 21,
              "textAlign": "center",
            }
          }
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry.
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly 3`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    collapsable={undefined}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 1334,
          },
        ],
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID={undefined}
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 10,
              "elevation": 24,
              "justifyContent": "center",
              "maxHeight": "70%",
              "maxWidth": "80%",
              "padding": 36,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 14.399999999999999,
              },
              "shadowOpacity": 0.216,
              "shadowRadius": 12.96,
            },
            Object {
              "marginBottom": NaN,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "height": 40,
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotate": "45deg",
                  },
                ],
                "width": 40,
              },
              Object {
                "borderBottomRightRadius": 3,
                "bottom": -20,
              },
            ]
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "backgroundColor": "transparent",
              "color": "#444",
              "fontSize": 21,
              "textAlign": "center",
            }
          }
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry.
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly 4`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    collapsable={undefined}
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    collapsable={undefined}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 1334,
          },
        ],
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID={undefined}
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 10,
              "elevation": 24,
              "justifyContent": "center",
              "maxHeight": "70%",
              "maxWidth": "80%",
              "padding": 36,
              "shadowColor": "black",
              "shadowOffset": Object {
                "height": 14.399999999999999,
              },
              "shadowOpacity": 0.216,
              "shadowRadius": 12.96,
            },
            Object {
              "marginRight": NaN,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "white",
                "height": 40,
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotate": "45deg",
                  },
                ],
                "width": 40,
              },
              Object {
                "borderTopRightRadius": 3,
                "right": -20,
              },
            ]
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "backgroundColor": "transparent",
              "color": "#444",
              "fontSize": 21,
              "textAlign": "center",
            }
          }
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry.
        </Text>
      </View>
    </View>
  </View>
</View>
`;
